File: java/src/tarehart/rlbot/planning/SetPieces.java
package tarehart.rlbot.planning;

import tarehart.rlbot.AgentOutput;
import tarehart.rlbot.math.vector.Vector2;
import tarehart.rlbot.steps.BlindStep;
import tarehart.rlbot.steps.TapStep;
import tarehart.rlbot.steps.landing.LandGracefullyStep;
import tarehart.rlbot.steps.landing.LandMindlesslyStep;
import tarehart.rlbot.steps.strikes.MidairStrikeStep;
import tarehart.rlbot.steps.travel.LineUpInReverseStep;
import tarehart.rlbot.time.Duration;
import tarehart.rlbot.tuning.ManeuverMath;

public class SetPieces {

    public static Plan frontFlip() {

        return new Plan()
                .unstoppable()
                .withStep(new BlindStep(.05,
                        new AgentOutput()
                                .withPitch(-1)
                                .withJump(true)
                                .withAcceleration(1)))
                .withStep(new BlindStep(.05,
                        new AgentOutput()
                                .withPitch(-1)
                                .withAcceleration(1)
                ))
                .withStep(new BlindStep(.3,
                        new AgentOutput()
                                .withJump(true)
                                .withAcceleration(1)
                                .withPitch(-1)))
                .withStep(new BlindStep(.5,
                        new AgentOutput()
                                .withAcceleration(1)
                                .withPitch(-1)
                ))
                .withStep(new LandGracefullyStep());
    }

    public static Plan halfFlip(Vector2 waypoint) {

        return new Plan()
                .unstoppable()
                .withStep(new LineUpInReverseStep(waypoint))
                .withStep(new BlindStep(.05,
                        new AgentOutput()
                                .withPitch(1)
                                .withJump(true)
                                .withAcceleration(-1)))
                .withStep(new BlindStep(.05,
                        new AgentOutput()
                                .withPitch(1)
                                .withAcceleration(-1)
                ))
                .withStep(new BlindStep(.05,
                        new AgentOutput()
                                .withJump(true)
                                .withPitch(1)))
                .withStep(new BlindStep(.15,
                        new AgentOutput()
                                .withPitch(1)))
                .withStep(new BlindStep(.4,
                        new AgentOutput()
                                .withBoost()
                                .withPitch(-1)
                ))
                .withStep(new BlindStep(.5,
                        new AgentOutput()
                                .withBoost()
                                .withPitch(-1)
                                .withSteer(1)
                                .withSlide()
                ));
    }

    public static Plan performAerial(double tiltBackSeconds) {
        Duration tiltBackDuration = Duration.Companion.ofSeconds(tiltBackSeconds);

        return new Plan()
                .withStep(new BlindStep(
                        new AgentOutput()
                                .withJump(true)
                                .withPitch(1),
                        tiltBackDuration
                ))
                .withStep(new MidairStrikeStep(tiltBackDuration))
                .withStep(new LandGracefullyStep(LandGracefullyStep.FACE_BALL));
    }

    public static Plan performDoubleJumpAerial(double tiltBackSeconds) {
        Duration tiltBackDuration = Duration.Companion.ofSeconds(tiltBackSeconds);

        return new Plan()
                .withStep(new BlindStep(
                        new AgentOutput()
                                .withJump(true)
                                .withPitch(1),
                        tiltBackDuration
                ))
                .withStep(new BlindStep(
                        new AgentOutput()
                                .withPitch(1)
                                .withBoost(true),
                        Duration.Companion.ofSeconds(.05)
                ))
                .withStep(new BlindStep(
                        new AgentOutput()
                                .withBoost(true)
                                .withJump(true),
                        Duration.Companion.ofSeconds(.05)
                ))
                .withStep(new MidairStrikeStep(tiltBackDuration))
                .withStep(new LandGracefullyStep(LandGracefullyStep.FACE_BALL));
    }

    public static Plan performJumpHit(double strikeHeight) {

        Double jumpSeconds = ManeuverMath.secondsForMashJumpHeight(strikeHeight).orElse(ManeuverMath.MASH_JUMP_HEIGHT);
        double pitchBackPortion = Math.min(.36, jumpSeconds);
        double driftUpPortion = jumpSeconds - pitchBackPortion;

        Plan plan = new Plan()
                .withStep(new BlindStep(
                        new AgentOutput()
                                .withJump(true)
                                .withPitch(1),
                        Duration.Companion.ofSeconds(pitchBackPortion)
                ));

        if (driftUpPortion > 0) {
            plan.withStep(new BlindStep(
                    new AgentOutput()
                            .withJump(true),
                    Duration.Companion.ofSeconds(driftUpPortion)
            ));
        }


        return plan
                .withStep(new TapStep(1,
                        new AgentOutput()
                                .withPitch(-1)
                                .withJump(false)
                                .withAcceleration(1)))
                .withStep(new TapStep(5,
                        new AgentOutput()
                                .withPitch(-1)
                                .withJump(true)
                                .withAcceleration(1)))
                .withStep(new BlindStep(
                        new AgentOutput()
                                .withAcceleration(1)
                                .withPitch(-1),
                        Duration.Companion.ofMillis(800)
                ))
                .withStep(new LandGracefullyStep(LandGracefullyStep.FACE_BALL));
    }

    public static Plan sideFlip(boolean flipLeft) {
        return new Plan()
                .unstoppable()
                .withStep(new TapStep(2,
                        new AgentOutput()
                                .withJump(true)
                                .withAcceleration(1)))
                .withStep(new TapStep(2,
                        new AgentOutput()
                                .withAcceleration(1)
                ))
                .withStep(new TapStep(2,
                        new AgentOutput()
                                .withJump(true)
                                .withAcceleration(1)
                                .withSteer(flipLeft ? -1 : 1)))
                .withStep(new LandMindlesslyStep());
    }

    public static Plan jumpSideFlip(boolean flipLeft, Duration jumpTime) {

        if (jumpTime.toMillis() < 0) {
            jumpTime = Duration.Companion.ofMillis(0);
        }

        return new Plan()
                .unstoppable()
                .withStep(new TapStep(2,
                        new AgentOutput()
                                .withJump(true)
                                .withAcceleration(1)))
                .withStep(new BlindStep(
                        new AgentOutput()
                                .withJump(true)
                                .withAcceleration(1), jumpTime))
                .withStep(new TapStep(2,
                        new AgentOutput()
                                .withAcceleration(1)
                ))
                .withStep(new TapStep(2,
                        new AgentOutput()
                                .withJump(true)
                                .withAcceleration(1)
                                .withSteer(flipLeft ? -1 : 1)))
                .withStep(new LandGracefullyStep(LandGracefullyStep.FACE_BALL));
    }

    public static Plan jumpSuperHigh(double howHigh) {
        return new Plan()
                .withStep(new BlindStep(
                        new AgentOutput()
                                .withJump(true)
                                .withPitch(1),
                        Duration.Companion.ofSeconds(.3)
                ))
                .withStep(new BlindStep(
                        new AgentOutput()
                                .withPitch(1)
                                .withBoost(true),
                        Duration.Companion.ofSeconds(.05)
                ))
                .withStep(new BlindStep(
                        new AgentOutput()
                                .withBoost(true)
                                .withJump(true),
                        Duration.Companion.ofSeconds(.05)
                ))
                .withStep(new BlindStep(
                        new AgentOutput()
                                .withJump(true)
                                .withBoost(true),
                        Duration.Companion.ofSeconds(howHigh / 10)
                ))
                .withStep(new LandGracefullyStep(LandGracefullyStep.FACE_BALL));
    }
}

File: java/src/tarehart/rlbot/planning/SetPieces.kt
package tarehart.rlbot.planning

import tarehart.rlbot.AgentOutput
import tarehart.rlbot.math.vector.Vector2
import tarehart.rlbot.steps.BlindStep
import tarehart.rlbot.steps.TapStep
import tarehart.rlbot.steps.landing.LandGracefullyStep
import tarehart.rlbot.steps.strikes.MidairStrikeStep
import tarehart.rlbot.steps.travel.LineUpInReverseStep
import tarehart.rlbot.time.Duration

object SetPieces {

    fun frontFlip(): Plan {

        return Plan()
                .unstoppable()
                .withStep(BlindStep(Duration.ofSeconds(.09),
                        AgentOutput()
                                .withPitch(-1.0)
                                .withJump(true)
                                .withThrottle(1.0)))
                .withStep(BlindStep(Duration.ofSeconds(.02),
                        AgentOutput()
                                .withPitch(-1.0)
                                .withThrottle(1.0)
                ))
                .withStep(BlindStep(Duration.ofSeconds(.3),
                        AgentOutput()
                                .withJump(true)
                                .withThrottle(1.0)
                                .withPitch(-1.0)))
                .withStep(BlindStep(Duration.ofSeconds(.5),
                        AgentOutput()
                                .withThrottle(1.0)
                                .withPitch(-1.0)
                ))
                .withStep(LandGracefullyStep({ it.myCarData.velocity.flatten()}))
    }

    fun halfFlip(waypoint: Vector2): Plan {

        return Plan()
                .unstoppable()
                .withStep(LineUpInReverseStep(waypoint))
                .withStep(BlindStep(Duration.ofSeconds(.05),
                        AgentOutput()
                                .withPitch(1.0)
                                .withJump(true)
                                .withThrottle(-1.0)))
                .withStep(BlindStep(Duration.ofSeconds(.05),
                        AgentOutput()
                                .withPitch(1.0)
                                .withThrottle(-1.0)
                ))
                .withStep(BlindStep(Duration.ofSeconds(.05),
                        AgentOutput()
                                .withJump(true)
                                .withPitch(1.0)))
                .withStep(BlindStep(Duration.ofSeconds(.15),
                        AgentOutput()
                                .withPitch(1.0)))
                .withStep(BlindStep(Duration.ofSeconds(.4),
                        AgentOutput()
                                .withBoost()
                                .withPitch(-1.0)
                ))
                .withStep(BlindStep(Duration.ofSeconds(.5),
                        AgentOutput()
                                .withBoost()
                                .withPitch(-1.0)
                                .withSteer(1.0)
                                .withSlide()
                ))
    }

    fun performAerial(tiltBackSeconds: Double): Plan {
        val tiltBackDuration = Duration.ofSeconds(tiltBackSeconds)

        return Plan()
                .withStep(BlindStep(tiltBackDuration, AgentOutput().withJump(true).withPitch(1.0)))
                .withStep(MidairStrikeStep(tiltBackDuration))
                .withStep(LandGracefullyStep(LandGracefullyStep.FACE_BALL))
    }

    fun performDoubleJumpAerial(tiltBackSeconds: Double): Plan {
        val tiltBackDuration = Duration.ofSeconds(tiltBackSeconds)

        return Plan()
                .withStep(BlindStep(tiltBackDuration, AgentOutput().withJump(true).withPitch(1.0)))
                .withStep(BlindStep(Duration.ofSeconds(.05), AgentOutput()
                        .withPitch(1.0)
                        .withBoost(true)
                ))
                .withStep(BlindStep(Duration.ofSeconds(.05), AgentOutput()
                        .withBoost(true)
                        .withJump(true)
                ))
                .withStep(MidairStrikeStep(tiltBackDuration))
                .withStep(LandGracefullyStep(LandGracefullyStep.FACE_BALL))
    }

    fun performJumpHit(jumpSeconds: Double): Plan {

        val pitchBackPortion = Math.min(.36, jumpSeconds)
        val driftUpPortion = jumpSeconds - pitchBackPortion

        val plan = Plan()
                .unstoppable()
                .withStep(BlindStep(Duration.ofSeconds(pitchBackPortion), AgentOutput()
                        .withJump(true)
                        .withPitch(1.0)
                ))

        if (driftUpPortion > 0) {
            plan.withStep(BlindStep(Duration.ofSeconds(driftUpPortion), AgentOutput().withJump(true)))
        }

        return plan
                .withStep(TapStep(1, AgentOutput()
                        .withPitch(-1.0)
                        .withJump(false)
                        .withThrottle(1.0)))
                .withStep(TapStep(5, AgentOutput()
                        .withPitch(-1.0)
                        .withJump(true)
                        .withThrottle(1.0)))
                .withStep(BlindStep(Duration.ofMillis(800), AgentOutput()
                        .withThrottle(1.0)
                        .withPitch(-1.0)
                ))
                .withStep(LandGracefullyStep(LandGracefullyStep.FACE_BALL))
    }

    fun jumpSideFlip(flipLeft: Boolean, rawJumpTime: Duration, hurry: Boolean): Plan {

        val jumpTime = if (rawJumpTime.millis < 0) Duration.ofMillis(0) else rawJumpTime

        return Plan()
                .unstoppable()
                .withStep(TapStep(2, AgentOutput()
                        .withJump(true)
                        .withThrottle((if (hurry) 1 else 0).toDouble())))
                .withStep(BlindStep(jumpTime, AgentOutput()
                        .withJump(true)
                        .withBoost(hurry)
                        .withThrottle(1.0)))
                .withStep(TapStep(2, AgentOutput()
                        .withThrottle(1.0)
                ))
                .withStep(TapStep(2, AgentOutput()
                        .withJump(true)
                        .withThrottle(1.0)
                        .withSteer((if (flipLeft) -1 else 1).toDouble())))
                .withStep(LandGracefullyStep(LandGracefullyStep.FACE_BALL))
    }

    fun jumpSuperHigh(howHigh: Double): Plan {
        return Plan()
                .withStep(BlindStep(Duration.ofSeconds(.3), AgentOutput()
                        .withJump(true)
                        .withPitch(1.0)
                ))
                .withStep(BlindStep(Duration.ofSeconds(.05), AgentOutput()
                        .withPitch(1.0)
                        .withBoost(true)
                ))
                .withStep(BlindStep(Duration.ofSeconds(.05), AgentOutput()
                        .withBoost(true)
                        .withJump(true)
                ))
                .withStep(BlindStep(Duration.ofSeconds(howHigh / 10), AgentOutput()
                        .withJump(true)
                        .withBoost(true)
                ))
                .withStep(LandGracefullyStep(LandGracefullyStep.FACE_BALL))
    }
}
